### PR TITLE
Add bzlmod support

### PR DESCRIPTION
**Pull request by aaronmondal**
Added a basic configuration for the new bzlmod module system which reduces the import of rules_m4 to a single line in a MODULE.bazel file.

### CODE REVIEW

**Comment by aaronmondal** [comment]
@jmillikin Hi, I'm working on integrating `rules_m4`, `rules_flex` and `rules_bison` into the BCR. The issue is tracked at https://github.com/bazelbuild/bazel-central-registry/issues/204

**Comment by aaronmondal** [comment]
Tests should also pass when the googletest dependency is pulled in via bzlmod. This can be tested by using a recent Bazel version such as 5.3.1 and prefixing the build and test commands with `--experimental_enable_bzlmod`. I not sure how to integrate that into the CI infra for this repo though.

**Review by jmillikin — COMMENTED** [review]
Is this code something that can be maintained in `bazel-central-registry`, or does it need to be present in the `rules_m4` repository? My understanding is that bzlmod is still experimental, so it might be best to have the module definition hosted somewhere with faster iteration cycles until the API stabilizes.

Separately,

I've had some discussions with the Bazel folks at BazelCon (pre-pandemic), and also left several comments on the bzlmod design doc (before it was locked down), but my concerns about the inflexibility of the "bazel federation" / "bzlmod" design were largely brushed aside. Specifically:
* Packages in the bzlmod design are no longer identified by URL, so while today this package is `github.com/jmillikin/rules_m4`, in the bzlmod world it would be just `rules_m4`.
  * I believe the experience of Go package management vs PyPI/NPM/Hackage/etc is strong evidence in favor of package namespacing. Even if full URLs are impractical, being able to identify this package as `~jmillikin/rules_m4` would avoid the sort of catastrophic namespace exhaustion that's affected NPM and crates.io.
  * If someone else had already created `rules_m4` before me, I wouldn't have been able to publish this package in the central registry. If I publish this, will it prevent someone else from publishing a better `rules_m4` in a year, or 5 years, or 10?
* In today's `WORKSPACE` model it's straightforward to have reliable zero-network builds in a Google-style monorepo through the use of the remote downloader and a `//third_party` directory containing dependency definition macros. The bzlmod design forbids the use of macros in `//:MODULE.bazel`, which means I'd either have to gain a hard dependency on the network, or maintain a file with thousands of `archive_override()` calls in it.
  * This is basically a hard-blocker to me using bzlmod for any of my own development, which means I would never be using or testing the module definitions of packages such as `rules_m4`.
  * Considering that Bazel is practically purpose-built for multi-project monorepos and bzlmod doesn't work for them, it's not clear to me that bzlmod has an audience large enough to justify its continued maintenance after the v1 has shipped. I don't want to have to migrate to it and then migrate back when it gets deprecated.

File: MODULE.bazel [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+    dev_dependency = True,
+)
+
+initialize_m4_toolchains = use_extension(
+    "@rules_m4//m4:extensions.bzl",
+    "initialize_m4_toolchains",
+)
+initialize_m4_toolchains.configure(extra_copts = [])
+use_repo(initialize_m4_toolchains, "m4_v1.4.18")
+register_toolchains("@rules_m4//m4/toolchains:v1.4.18")
```
Comment (jmillikin) [inline MODULE.bazel]: I'm not familiar enough with the bzlmod "extensions" API to say for sure, but would the following code also work?
```python
m4 = use_extension("@rules_m4//m4:extensions.bzl", "m4_extension")
m4.register_toolchains()  # using default options
m4.register_toolchains(version = "1.4.17", extra_copts = ["-w"])
```

Where the extension looks something like this:
```python
def _m4_extension(module_ctx):
    for module in module_ctx.modules:
        for register in module.tags.register_toolchains:
            check_version(register.version)
            repo_name = "m4_v{}".format(register.version)
            m4_repository(
                name = repo_name,
                version = register.version,
                extra_copts = register.extra_copts,
            )

            native.register_toolchains("@rules_m4//m4/toolchains:v{}".format(version))
```

**Comment by aaronmondal** [comment]
Thanks for the review :blush: I'll try my best to answer your questions, but I should note that I am not affiliated with the Bazel team or Google :sweat_smile: We used the m4 etc rules in our own bazel registry and are now moving that registry closer to the BCR so we thought it would be a good idea to clean up our bzlmod overlays and upstream them.

> Is this code something that can be maintained in `bazel-central-registry`, or does it need to be present in the `rules_m4` repository? My understanding is that bzlmod is still experimental, so it might be best to have the module definition hosted somewhere with faster iteration cycles until the API stabilizes.

One could apply patches to a repo on the BCR side if it does not have bzlmod support. So essentially the `git diff` for this PR would be stored in the BCR. This would not require changing `rules_m4`, but makes things less intuitive to work with on the BCR side since these patches need to be updated for every new release. If `rules_m4` natively supports bzlmod, the BCR only needs to store a copy of MODULE.bazel. If `--experimental_enable_bzlmod` is not flipped, the `MODULE.bazel` and `extensions.bzl` files will essentially be completely ignored and the WORKSPACE workflow is unaffected.

> My understanding is that bzlmod is still experimental, so it might be best to have the module definition hosted somewhere with faster iteration cycles until the API stabilizes.

Regarding API changes, from what I've witnessed so far in the last ~year was only one API change, (moving the `register_toolchains` out of the `module` macro - a 2 line change). Everything else were bugfixes. I would expect that there will be a similar change around `register_platforms`, but so far I've experienced the bzlmod API as pretty much stable. The BCR though does not appear as stable to me. I think it is currently lacking proper guidelines on how and when to add modules. I think bzlmod support should be seen as distinct from being tracked in the BCR. For instance, we are not actually using the BCR itself in our builds. We have our own vendored bazel registry (which already had `rules_{m4,flex,bison}` modules), but we are transitioning it to become a fork of the BCR. We thought it would be nice to upstream some of the modules we have into the BCR to  make them available to a broader audience.

> * Packages in the bzlmod design are no longer identified by URL, so while today this package is `github.com/jmillikin/rules_m4`, in the bzlmod world it would be just `rules_m4`.

Technically, the URLs are still tracked (in the bazel registries in `source.json` files) and can be arbitrarily overridden. 

We've actually run into these naming issues. We have our own overlay for `abseil` that uses completely different build files. So there is kind of like an `absl/abseil-cpp` and an `eomii/abseil-cpp`. In our old registry, there was no `absl/abseil-cpp` and  `eomii/abseil-cpp` was simply called `abseil-cpp`. At some point maintaining the transitive dependencies that were not our own in our registry became too tedious. Our solution for our new BCR-fork registry is to prefix our own dependencies to differentiate them. For instance, we will have `abseil-cpp` from the BCR and `ll_abseil-cpp` for our own version.

For vendored registries I think there is no issue with prefixing like that, but I absolutely see the issue for users that depend only on the "official" BCR. One would expect the "most official" ruleset for `something` to be called `rules_something`. Being forced to use a different name may not always desirable (I'm working on `rules_ll` which is an orthogonal approach to `rules_cc`. It would've been very confusing to also name our rules `rules_cc`. In our case we wanted a different name, but for many rulesets naming schemes could become an issue.)

>   * I believe the experience of Go package management vs PyPI/NPM/Hackage/etc is strong evidence in favor of package namespacing. Even if full URLs are impractical, being able to identify this package as `~jmillikin/rules_m4` would avoid the sort of catastrophic namespace exhaustion that's affected NPM and crates.io.
>   * If someone else had already created `rules_m4` before me, I wouldn't have been able to publish this package in the central registry. If I publish this, will it prevent someone else from publishing a better `rules_m4` in a year, or 5 years, or 10?

I don't think that it is impossible to change the BCR to use names like that at a later point in time. A module can be named `my_module` in the local MODULE.bazel and still be renamed to `aaron.my_module` or similar in the MODULE.bazel that is stored in the BCR. If one were to make such a change when people already depend on `my_module`, builds would break for downstream users, but not in the `my_module` repo. As long as the fix for downstream users is to just add an `aaron.` to the module name I think such a change is acceptable. I think I'll bring up these namespace concerns in the BCR repo though. I think this is fixable without any actual API changes (and without changing all the downstream repos), but it needs to be talked about. There should be proper guidelines around this topic.

> * In today's `WORKSPACE` model it's straightforward to have reliable zero-network builds in a Google-style monorepo through the use of the remote downloader and a `//third_party` directory containing dependency definition macros. The bzlmod design forbids the use of macros in `//:MODULE.bazel`, which means I'd either have to gain a hard dependency on the network, or maintain a file with thousands of `archive_override()` calls in it.
>   * This is basically a hard-blocker to me using bzlmod for any of my own development, which means I would never be using or testing the module definitions of packages such as `rules_m4`.

I've not yet played around with true offline builds. We mostly override specific dependencies with local repositories during development. As a naive thought, I think it should be reasonably simple to create a script that parses a bazel registry, downloads all dependencies once into a monorepo and then rewrites the URLs in the registry to point to that local monorepo. Then the modified registry is stored in the monorepo as well and we should have something similar to the WORKSPACE offline workflow. Something like this may actually be more convenient to work with than the WORKSPACE workflow, since one only needs to point bazel to the modified registry in the `.bazelrc`, and all other build files can remain unchanged. And if one decides to switch to e.g. an online registry, only the registry URL in a single `.bazelrc` file needs to change.

>   * Considering that Bazel is practically purpose-built for multi-project monorepos and bzlmod doesn't work for them, it's not clear to me that bzlmod has an audience large enough to justify its continued maintenance after the v1 has shipped. I don't want to have to migrate to it and then migrate back when it gets deprecated.

I think bzlmod still enables monorepo workflows, but instead of WORKSPACE files referencing local archives, one would roll their own offline bazel registry and point Bazel to that offline registry.

As far as I understood, bzlmod will deprecate the WORKSPACE workflow in 6.0 or 7.0. I am very biased here, but I'm fairly confident that bzlmod will stay. We've been using it from pretty much the moment it came out. It is really nice to work with since all the WORKSPACE logic just vanishes and almost every module import becomes a single LoC. Keeping things up-to-date only requires bumping a version number instead of checking for hashes. For reference, this is our MODULE.bazel for `rules_ll`: https://github.com/eomii/rules_ll/blob/main/MODULE.bazel. (And this was our old bazel registry which we are now transitioning: https://github.com/eomii/bazel-eomii-registry-deprecated/tree/main/modules. I could imagine that changing all the URLs in the `source.json` files to local URLs should be enough to set something like this up inside a monorepo.)

File: MODULE.bazel (hunk 1 above)
Comment (aaronmondal) [reply MODULE.bazel]: I'll get back to you on this one. I'll check tomorrow.

**Comment by aaronmondal** [comment]
@jmillikin This new variant supports configurations (and usage) of specific m4 versions. All toolchains are registered in the MODULE.bazel for rules_m4, so users will not need to register toolchains themselves. So a user using the default toolchain configuration would use
```python
bazel_dep(name = "rules_m4", version = "0.2.2-bcr.0")
```
and toolchains are automatically registered. If a user wanted to reconfigure a specific toolchain, the following would do that:
```python
bazel_dep(name = "rules_m4", version = "0.2.2-bcr.0")
m4 = use_extension("@rules_m4//m4:extensions.bzl", "m4_extension")
m4.configure_toolchain(version = "1.4.18", extra_copts = ["-w"])
```
The new version of this PR makes it so that using `m4.configure` effectively overrides the `extra_copts` for that specific toolchain build.

It is not possible to call `native.register_toolchains` from a module extension, but having an `m4.register_toolchain` would kinda go against the "standard" way of registering toolchains in bzlmod anyways.

A potential question would be whether all these toolchains should be registered automatically. I think the current variant is nice since it lets users use `rules_m4` with a single import LoC. An alternative would be to not register the toolchains in `rules_m4`, but then users would have to do something like
```python
bazel_dep(name = "rules_m4", version = "0.2.2-bcr.0")
m4 = use_extension("@rules_m4//m4:extensions.bzl", "m4_extension")
m4.configure_toolchain(version = "1.4.18")
use_repo(m4, "v1.4.18")
register_toolchains("@rules_m4//m4/toolchains:v1.4.18")
```
which I think is awfully verbose if we can achieve essentially the same thing with the current variant (also, this code also still works with the current implementation, so if someone would prefer this way, it still works).

**Comment by jmillikin** [comment]
This might be because the Bazel I'm using is too old (v5.3), or my own lack of experience with bzlmod, but I haven't been able to get selection of a specific version to work as documented.

First attempt, with `rules_m4/MODULE.bazel` registering all toolchains, and the test repo requesting (?) M4 v1.4.17:
```python
# MODULE.bazel
module()
bazel_dep(
    name = "rules_m4",
    version = "0.2.2",
)
local_path_override(
    module_name = "rules_m4",
    path = "/path/to/rules_m4/pr12",
)
m4 = use_extension("@rules_m4//m4:extensions.bzl", "m4_extension")
m4.configure_toolchain(version = "1.4.17")
```
```python
# BUILD
load("@rules_m4//m4:m4.bzl", "m4")
m4(
    name = "m4_version",
    srcs = ["empty.txt"],
    m4_options = ["--version"],
    output = "m4_version.txt",
)
```
```
$ bazel build --experimental_enable_bzlmod //:m4_version
[...]
Target //:m4_version up-to-date:
  bazel-bin/m4_version.txt
[...]
$ head -n1 bazel-bin/m4_version.txt
m4 (GNU M4) 1.4.18
```

Second attempt, commenting out the toolchain registration in `rules_m4/MODULE.bazel` and adding it to the test repo (the "awfully verbose" example from your last comment):
```python
# MODULE.bazel
module()
bazel_dep(
    name = "rules_m4",
    version = "0.2.2",
)
local_path_override(
    module_name = "rules_m4",
    path = "/home/john/src/third_party/github.com/jmillikin/rules_m4/trunk",
)
m4 = use_extension("@rules_m4//m4:extensions.bzl", "m4_extension")
m4.configure_toolchain(version = "1.4.17")
use_repo(m4, "m4_v1.4.17")
register_toolchains("@rules_m4//m4/toolchains:v1.4.17")
```
```
[...]
ERROR: /home/john/.cache/bazel/_bazel_john/29242636fa50cd73631461833f385049/external/rules_m4.override/m4/internal/BUILD:4:19: no such package '@m4_v1.4.17//bin': The repository '@m4_v1.4.17' could not be resolved: Repository '@m4_v1.4.17' is not visible from repository '@rules_m4.override' and referenced by '@rules_m4.override//m4/internal:toolchain_v1.4.17'
ERROR: Analysis of target '//:m4_version' failed; build aborted: 
```

This is presumably the bzlmod "strict deps" behavior in action, but I would have expected `@rules_m4` to have access to `@m4_v1.4.17` via the `use_repo()` call:
> A module extension repo can see all visible dependencies of the module that provides the extension, **plus all other repos generated by the same module extension**.

The only way I could get it to work is to have `rules_m4` declare all the versions, but then have the test repo register the toolchain for the version it wants. This isn't great at present because the M4 toolchain targets are per-version, but maybe there's some way to have a "default version toolchain" as an alias.

I'll do some more poking at this to get a better understanding of how bzlmod works.

**Comment by jmillikin** [comment]
@aaronmondal Bazel v6.2 (released 3 days ago) included some bzlmod features that unblocked this work. I've pushed a dev branch to https://github.com/jmillikin/rules_m4/tree/with-bzlmod -- could you try it out and let me know if it works as you expect?

**Basic instructions**

Adding just the dependency will register an M4 toolchain using the default version (v1.4.18). This should be all that's needed for the common case.

```python
bazel_dep(name = "rules_m4", version = "0.2.3")
```

To add additional versions:

```python
m4 = use_extension("@rules_m4//m4/extensions:m4_repository.bzl", "m4_repository")

# Repository names default to `m4_v{version}`.
#
# Calling `use_repo()` will enable direct
# invocation: `bazel run @m4_v1.4.17//bin:m4 -- --version`, etc.
m4.repository(version = "1.4.17")
use_repo(m4, "m4_v1.4.17")

# The repository name can be adjusted
m4.repository(name = "m4_older", version = "1.4.17")
use_repo(m4, "m4_older")

# Also supports `extra_copts`
m4.repository(
    name = "m4_fast",
    version = "1.4.18",
    extra_copts = ["-O3"],
)
use_repo(m4, "m4_fast")
```

To register a toolchain for a non-default version, use `register_toolchains`. This version will be used in preference to the default version registered by `rules_m4`.

```python
m4.repository(version = "1.4.17")
use_repo(m4, "m4_v1.4.17")
register_toolchains("@m4_v1.4.17//:toolchain")
```

**Comment by jmillikin** [comment]
I've polished up the dev branch linked above and merged it in commit 31ef50b18073a537bb085043faa112163dec4c9a.

Release v0.2.3 includes `MODULE.bazel`. I've filed a PR at https://github.com/bazelbuild/bazel-central-registry/pull/633 to get it into BCR.

**Pull request closed without merging**